cpp/src/arrow/util/cache_internal.h

```diff
@@ -139,7 +137,7 @@ struct ThreadSafeMemoizer {
   }
 
  private:
-  std::unique_ptr<std::mutex> mutex_;
```

Comment:
The `unique_ptr` was meant to make the memoizer type movable. While moving isn't strictly required, non-movable types are generally annoying to use (especially as some compilers may not always implement return value optimization properly).